packages/hebao_v1/contracts/modules/core/ForwarderModule.sol

```diff
@@ -141,23 +144,21 @@ abstract contract ForwarderModule is BaseModule
             controller().nonceStore().verifyAndUpdate(metaTx.from, metaTx.nonce);
         }
 
-        uint gasUsed = gasLeft - gasleft();
-
-        emit MetaTxExecuted(
-            msg.sender,
-            metaTx.from,
-            metaTx.nonce,
-            metaTx.txAwareHash,
-            success,
-            gasUsed
-        );
-
         // Fees are not to be charged by a relayer if the transaction fails with a
         // non-zero txAwareHash. The reason is that relayer can pick arbitrary 'data'
         // to make the transaction fail. Charging fees for such failures can drain
         // wallet funds.
-        if (metaTx.gasPrice > 0 && (metaTx.txAwareHash == 0 || success)) {
-            uint gasAmount = gasUsed < metaTx.gasLimit ? gasUsed : metaTx.gasLimit;
+        bool needReimburst = metaTx.gasPrice > 0 && (metaTx.txAwareHash == 0 || success);
+
+        uint gasUsed = gasLeft - gasleft() +
+            (signature.length + metaTx.data.length + 7) * 16 + // data input cost
+            447 + // cost of MetaTxExecuted = 375 + 9 * 8
+            21000; // transaction cost;
+
+        if (needReimburst) {
+            gasUsed = gasUsed +
+                MAX_REIMBURSTMENT_OVERHEAD + // near-worst case cost
+                1500; // 3 * 500 for three SLOAD for calculating skipQuota below
```

Comment:
SLOAD cost should be [800 gas](https://github.com/sorpaas/EIPs/blob/c88b73889e66669db0e7094dda7907661f54d873/EIPS/eip-2200.md#specification) I think.  If this is for `controller().walletFactory()` I think the cost is `2*SLOAD+1*CALL = 2*800+1*700=2300` but I don't really know how accurate that will be either.